File: python/examples/ex08_machines.py
"""Example 08 -- Machines lifecycle with cost-awareness and guaranteed cleanup.

Purpose
    Provision a cloud VM, drive it (screenshot, low-level actions, a batch,
    terminal, files, browser), snapshot it, and ALWAYS stop + terminate it in
    a ``finally`` block -- printing a running cost estimate the whole way.

Flow
    spend gate -> POST /v1/machines (ttl_minutes auto-terminate guard)
    -> PATCH /v1/machines/{id} (re-arm the TTL spend guard)
    -> poll GET /v1/machines/{id} until running
    -> GET /v1/machines/{id}/screenshot -> save PNG
    -> POST .../actions (click, type_text) -> POST .../actions/batch
    -> POST .../terminal -> POST .../files/write + .../files/read
    -> POST .../browser/navigate -> POST .../snapshot ($0.01)
    -> finally: POST .../stop + DELETE /v1/machines/{id}

Endpoints
    POST/GET/PATCH/DELETE /v1/machines[/{id}], /screenshot, /actions,
    /actions/batch, /terminal, /files/{op}, /browser/{op}, /snapshot, /stop

Estimated cost (via coasty.cost, printed before provisioning)
    Runtime is hourly, metered per minute, rounded down: Linux 5 cr/hr,
    Windows 9 cr/hr while running; 1 cr/hr stopped; creating/terminated free.
    Snapshot: 1 cr ($0.01), refunded on failure. Per-call ops (actions,
    terminal, files, browser, screenshots, start/stop) are FREE.
    Sandbox keys get an instant free ``mch_test_*`` VM ($0); live keys need a
    wallet of at least 20 cr ($0.20) to provision (a gate, not a fee) and
    must pass --confirm / COASTY_CONFIRM_SPEND=1.

Run
    python examples/ex08_machines.py --os linux --ttl-minutes 30 \
        --screenshot ex08_screenshot.png --confirm
"""

from __future__ import annotations

import argparse
import base64
import sys
import time
from collections.abc import Callable, Sequence
from dataclasses import dataclass
from pathlib import Path
from typing import Any

from coasty import CoastyClient, CoastyError, cost, env
from coasty.types import OsType

DEFAULT_DISPLAY_NAME = "cookbook-ex08"
DEFAULT_TTL_MINUTES = 30
DEFAULT_BROWSER_URL = "https://example.com"
REMOTE_FILE_PATH = "/tmp/coasty-ex08.txt"
REMOTE_FILE_CONTENT = "hello from the coasty cookbook (ex08)"
PROVISION_WALLET_GATE_CREDITS = 20


class SpendNotConfirmedError(RuntimeError):
    """Raised when a billable lifecycle is attempted on a live key w/o consent."""


class MachineNeverBecameReadyError(RuntimeError):
    """Polling gave up before the machine reached ``running``."""

    def __init__(self, machine_id: str, status: str, request_id: str | None) -> None:
        super().__init__(
            f"machine {machine_id} never reached 'running' (last status {status!r}, "
            f"request_id={request_id})"
        )
        self.machine_id = machine_id
        self.last_status = status
        self.request_id = request_id


@dataclass(frozen=True)
class LifecycleReport:
    """Summary of one full machine lifecycle, for printing and for tests."""

    machine_id: str
    os_type: OsType
    is_test_machine: bool
    provision_request_id: str | None
    snapshot_id: str | None
    screenshot_path: str | None
    terminal_excerpt: str | None
    file_roundtrip_ok: bool
    elapsed_minutes: float
    estimated_runtime_credits: int
    stopped: bool
    terminated: bool


def ensure_provision_allowed(
    estimate: cost.CostEstimate,
    *,
    sandbox: bool,
    os_type: OsType,
    confirm: bool,
    printer: Callable[[str], None] = print,
) -> None:
    """Print the itemized estimate + live-key gates; refuse without consent."""
    rate = cost.machine_hourly_credits(os_type, "running")
    printer(cost.format_estimate(estimate, title="Estimated machine cost", sandbox=sandbox))
    printer(f"hourly rate while running ({os_type}): {rate} cr/hr, metered per minute")
    if sandbox:
        printer("sandbox key: provisioning is instant and free (mch_test_* VM, $0)")
        return
    printer(
        f"live key: provisioning requires a wallet balance of >= "
        f"{PROVISION_WALLET_GATE_CREDITS} cr ($0.20) -- a gate, not a fee"
    )
    if confirm:
        return
    raise SpendNotConfirmedError(
        "refusing to provision on a live key: pass --confirm or set COASTY_CONFIRM_SPEND=1 "
        "(or use an sk-coasty-test-... sandbox key for a free instant VM)"
    )


def _wait_until_running(
    client: CoastyClient,
    machine_id: str,
    initial_status: str,
    *,
    poll_interval: float,
    max_polls: int,
    sleep: Callable[[float], None],
    printer: Callable[[str], None],
) -> None:
    """Poll GET /v1/machines/{id} until status == running (or give up loudly)."""
    status = initial_status
    last_request_id: str | None = None
    for _ in range(max_polls):
        if status == "running":
            return
        sleep(poll_interval)
        polled = client.get_machine(machine_id)
        machine_raw = polled.data.get("machine", polled.data)
        machine: dict[str, Any] = machine_raw if isinstance(machine_raw, dict) else {}
        status = str(machine.get("status", "unknown"))
        last_request_id = polled.request_id
        printer(f"machine {machine_id}: status={status} (request_id={polled.request_id})")
    if status != "running":
        raise MachineNeverBecameReadyError(machine_id, status, last_request_id)


def run_lifecycle(
    client: CoastyClient,
    *,
    display_name: str = DEFAULT_DISPLAY_NAME,
    os_type: OsType = "linux",
    ttl_minutes: int = DEFAULT_TTL_MINUTES,
    screenshot_path: Path | None = None,
    browser_url: str = DEFAULT_BROWSER_URL,
    confirm_spend: bool = False,
    poll_interval: float = 2.0,
    max_polls: int = 30,
    clock: Callable[[], float] = time.monotonic,
    sleep: Callable[[float], None] = time.sleep,
    printer: Callable[[str], None] = print,
) -> LifecycleReport:
    """Provision, exercise, snapshot, and ALWAYS stop + terminate the machine.

    The spend gate runs before any network call; tests can verify a live key
    without ``confirm_spend`` never even provisions.
    """
    estimate = cost.combine(
        cost.estimate_machine_runtime(os_type=os_type, state="running", minutes=ttl_minutes),
        cost.estimate_snapshot(),
    )
    ensure_provision_allowed(
        estimate,
        sandbox=client.is_sandbox,
        os_type=os_type,
        confirm=confirm_spend,
        printer=printer,
    )

    provisioned = client.provision_machine(
        display_name,
        os_type=os_type,
        ttl_minutes=ttl_minutes,  # auto-terminate: the machine cannot outlive this guard
    )
    machine = provisioned.data["machine"]
    machine_id = machine["id"]
    printer(
        f"provisioned {machine_id} ({machine['os_type']}, status={machine['status']}, "
        f"is_test={machine['is_test']}) ttl={ttl_minutes}min "
        f"(request_id={provisioned.request_id})"
    )
    started_at = clock()

    snapshot_id: str | None = None
    saved_screenshot: str | None = None
    terminal_excerpt: str | None = None
    file_roundtrip_ok = False
    stopped = False
    terminated = False
    try:
        # Re-arm the TTL explicitly: PATCH is how you extend/shrink the spend
        # guard on a machine that already exists (0 clears it -- never do that
        # in a cost-aware script).
        ttl_ack = client.set_machine_ttl(machine_id, ttl_minutes)
        printer(f"ttl re-armed at {ttl_minutes}min (request_id={ttl_ack.request_id})")

        _wait_until_running(
            client,
            machine_id,
            str(machine["status"]),
            poll_interval=poll_interval,
            max_polls=max_polls,
            sleep=sleep,
            printer=printer,
        )

        shot = client.machine_screenshot(machine_id)
        if screenshot_path is not None:
            screenshot_path.write_bytes(base64.b64decode(shot.data["image_b64"]))
            saved_screenshot = str(screenshot_path)
            printer(
                f"screenshot {shot.data['width']}x{shot.data['height']} saved to "
                f"{screenshot_path} (request_id={shot.request_id})"
            )

        click = client.machine_action(machine_id, "click", parameters={"x": 640, "y": 360})
        printer(f"action click: success={click.data['success']} (request_id={click.request_id})")
        typed = client.machine_action(
            machine_id, "type_text", parameters={"text": REMOTE_FILE_CONTENT}
        )
        printer(f"action type_text: success={typed.data['success']}")

        batch = client.machine_actions_batch(
            machine_id,
            steps=[
                {"command": "key_combo", "parameters": {"keys": ["ctrl", "a"]}},
                {"command": "wait", "parameters": {"ms": 250}},
                {"command": "key_press", "parameters": {"key": "escape"}},
            ],
            stop_on_error=True,
        )
        printer(
            f"batch: {batch.data['completed_count']} completed, "
            f"{batch.data['failed_count']} failed, aborted={batch.data['aborted']} "
            f"(request_id={batch.request_id})"
        )

        term = client.machine_terminal(machine_id, "echo coasty-cookbook && uname -a")
        terminal_output = term.data.get("output") or term.data.get("result") or ""
        terminal_excerpt = str(terminal_output)[:200]
        printer(f"terminal: {terminal_excerpt!r} (request_id={term.request_id})")

        client.machine_files(
            machine_id,
            "write",
            {"path": REMOTE_FILE_PATH, "content": REMOTE_FILE_CONTENT},
        )
        read_back = client.machine_files(machine_id, "read", {"path": REMOTE_FILE_PATH})
        # The docs do not pin the files-op response shape: some servers return
        # the content at the top level, others nest it under "result" like the
        # documented /actions envelope. Accept both.
        nested = read_back.data.get("result")
        read_content = nested.get("content") if isinstance(nested, dict) else None
        if read_content is None:
            read_content = read_back.data.get("content")
        file_roundtrip_ok = read_content == REMOTE_FILE_CONTENT
        printer(f"file write+read roundtrip ok={file_roundtrip_ok}")

        nav = client.machine_browser(machine_id, "navigate", parameters={"url": browser_url})
        printer(f"browser navigate -> {browser_url} (request_id={nav.request_id})")

        snap = client.snapshot_machine(machine_id)
        snapshot_id = snap.data["snapshot_id"]
        printer(
            f"snapshot {snapshot_id} taken -- 1 cr ($0.01), refunded on failure "
            f"(request_id={snap.request_id})"
        )
    finally:
        # Stop first (drops billing to 1 cr/hr) then terminate (billing ends).
        # Cleanup failures are reported, never silently swallowed, and never
        # mask the original in-flight exception.
        try:
            stop_ack = client.stop_machine(machine_id)
            stopped = True
            printer(f"stopped {machine_id} (request_id={stop_ack.request_id})")
        except CoastyError as exc:
            printer(f"WARNING: stop failed: {exc}")
        try:
            term_ack = client.terminate_machine(machine_id)
            terminated = True
            printer(f"terminated {machine_id} (request_id={term_ack.request_id})")
        except CoastyError as exc:
            printer(f"WARNING: terminate failed: {exc} -- the ttl_minutes guard still applies")

    elapsed_minutes = max(0.0, (clock() - started_at) / 60.0)
    runtime = cost.estimate_machine_runtime(
        os_type=os_type, state="running", minutes=elapsed_minutes
    )
    printer(cost.format_estimate(runtime, title="Actual runtime cost (running-state rate)"))
    return LifecycleReport(
        machine_id=machine_id,
        os_type=os_type,
        is_test_machine=bool(machine["is_test"]),
        provision_request_id=provisioned.request_id,
        snapshot_id=snapshot_id,
        screenshot_path=saved_screenshot,
        terminal_excerpt=terminal_excerpt,
        file_roundtrip_ok=file_roundtrip_ok,
        elapsed_minutes=elapsed_minutes,
        estimated_runtime_credits=runtime.credits,
        stopped=stopped,
        terminated=terminated,
    )


# ── CLI ────────────────────────────────────────────────────────────────────


def main(argv: Sequence[str] | None = None) -> int:
    parser = argparse.ArgumentParser(description=__doc__.split("\n", maxsplit=1)[0])
    parser.add_argument("--display-name", default=DEFAULT_DISPLAY_NAME)
    parser.add_argument("--os", dest="os_type", choices=("linux", "windows"), default="linux")
    parser.add_argument("--ttl-minutes", type=int, default=DEFAULT_TTL_MINUTES)
    parser.add_argument("--screenshot", type=Path, default=Path("ex08_screenshot.png"))
    parser.add_argument("--browser-url", default=DEFAULT_BROWSER_URL)
    parser.add_argument("--poll-interval", type=float, default=2.0)
    parser.add_argument("--max-polls", type=int, default=30)
    parser.add_argument("--confirm", action="store_true", help="consent to spend on a live key")
    args = parser.parse_args(argv)

    api_key = env.require_api_key()
    os_type: OsType = "windows" if args.os_type == "windows" else "linux"
    try:
        with CoastyClient(api_key=api_key) as client:
            report = run_lifecycle(
                client,
                display_name=args.display_name,
                os_type=os_type,
                ttl_minutes=args.ttl_minutes,
                screenshot_path=args.screenshot,
                browser_url=args.browser_url,
                confirm_spend=args.confirm or env.spend_confirmed(),
                poll_interval=args.poll_interval,
                max_polls=args.max_polls,
            )
    except SpendNotConfirmedError as exc:
        print(f"error: {exc}", file=sys.stderr)
        return 2
    except CoastyError as exc:
        print(
            f"API error {exc.code} (request_id={exc.request_id}): {exc.message}",
            file=sys.stderr,
        )
        return 1
    print(
        f"done: machine={report.machine_id} snapshot={report.snapshot_id} "
        f"stopped={report.stopped} terminated={report.terminated} "
        f"runtime~{report.elapsed_minutes:.1f}min (~{report.estimated_runtime_credits} cr)"
    )
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: typescript/src/examples/ex08-machines.ts
/**
 * Example 08 — Machines lifecycle, cost-aware.
 *
 * provision -> poll until running -> screenshot (saved to disk) -> one
 * low-level action -> a batched action sequence -> terminal command -> file
 * write/read -> browser navigate -> snapshot -> STOP + TERMINATE in finally.
 *
 * Cost model (computed for the defaults: Linux, 30-minute TTL):
 *   - provisioning gate: wallet must hold >= 20 credits ($0.20) — a GATE, not
 *     a fee; sandbox keys skip it entirely (instant free mch_test_* VM);
 *   - runtime: Linux 5 cr/hr running (Windows 9 cr/hr), 1 cr/hr stopped,
 *     metered per minute and ROUNDED DOWN -> a full 30-minute session costs
 *     floor(5 x 30 / 60) = 2 credits ($0.02);
 *   - snapshot: 1 credit ($0.01) one-time, refunded on failure;
 *   - actions / batch / terminal / files / browser / screenshots / stop:
 *     0 credits per call.
 *   TOTAL (typical run, well under TTL): <= 3 credits ($0.03); $0.00 on a
 *   sandbox key. A wallet-dry machine is STOPPED, never destroyed.
 *
 * The `ttl_minutes` guard is mandatory in this example: every provision sets
 * an auto-terminate TTL (5-10080 min) so a crashed script can never leak a
 * billing machine.
 *
 * Run it:
 *   npx tsx src/examples/ex08-machines.ts [--os linux|windows] [--ttl 30] \
 *     [--out ./ex08-screenshot.png] [--confirm]
 */
import { writeFile } from 'node:fs/promises';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

import { CoastyClient } from '../coasty/client.js';
import {
  PRICING,
  creditsToUsd,
  estimateMachineRuntimeCredits,
  formatEstimate,
  machineHourlyCredits,
  snapshotCredits,
} from '../coasty/cost.js';
import { isSandboxKey, spendConfirmed } from '../coasty/env.js';
import { CoastyError } from '../coasty/errors.js';
import { defaultSleep, type SleepFn } from '../coasty/sse.js';
import { type Machine, type OsType } from '../coasty/types.js';

// ---------------------------------------------------------------------------
// Guards
// ---------------------------------------------------------------------------

export const TTL_MIN_MINUTES = 5;
export const TTL_MAX_MINUTES = 10080;
export const DEFAULT_TTL_MINUTES = 30;

/** Thrown when a live (billable) key is used without explicit confirmation. */
export class SpendNotConfirmedError extends Error {
  constructor(osType: OsType) {
    const hourly = machineHourlyCredits(osType, 'running');
    super(
      `Refusing to provision a live ${osType} machine (${String(hourly)} cr/hr running, ` +
        `gate: wallet >= ${String(PRICING.provisioningGateCredits)} cr) without confirmation. ` +
        'Pass --confirm (or set COASTY_CONFIRM_SPEND=1), or use an sk-coasty-test-* sandbox key.',
    );
    this.name = 'SpendNotConfirmedError';
  }
}

/** Thrown when the mandatory auto-terminate TTL is missing or out of range. */
export class TtlGuardError extends Error {
  constructor(ttlMinutes: number) {
    super(
      `ttl_minutes must be an integer between ${String(TTL_MIN_MINUTES)} and ` +
        `${String(TTL_MAX_MINUTES)} (got ${String(ttlMinutes)}). This example refuses to ` +
        'provision a machine without an auto-terminate TTL — a crashed script must never ' +
        'leak a billing VM.',
    );
    this.name = 'TtlGuardError';
  }
}

/** Thrown when a per-call machine op reports success=false (never silent). */
export class MachineOpFailedError extends Error {
  readonly requestId: string | null;

  constructor(op: string, detail: string | null, requestId: string | null) {
    super(
      `machine op "${op}" failed: ${detail ?? 'no error detail'} (request_id: ${requestId ?? 'n/a'})`,
    );
    this.name = 'MachineOpFailedError';
    this.requestId = requestId;
  }
}

export function validateTtlMinutes(ttlMinutes: number): void {
  if (
    !Number.isInteger(ttlMinutes) ||
    ttlMinutes < TTL_MIN_MINUTES ||
    ttlMinutes > TTL_MAX_MINUTES
  ) {
    throw new TtlGuardError(ttlMinutes);
  }
}

// ---------------------------------------------------------------------------
// Testable core
// ---------------------------------------------------------------------------

export interface MachineExampleOptions {
  /** True when the API key is a sandbox key (instant free mch_test_* VM). */
  sandbox: boolean;
  /** True when the user explicitly confirmed live spend. */
  confirmedSpend: boolean;
  /** Where to save the screenshot PNG. */
  screenshotPath: string;
  displayName?: string;
  osType?: OsType;
  /** Auto-terminate TTL (5-10080); ALWAYS set. Default 30. */
  ttlMinutes?: number;
  /** Provision/snapshot retries become safe with this. */
  idempotencyKey?: string;
  pollIntervalMs?: number;
  maxPolls?: number;
  sleep?: SleepFn;
  /** Millisecond clock for the runtime cost readout (injectable for tests). */
  now?: () => number;
  logger?: (line: string) => void;
}

export interface MachineExampleReport {
  machine: Machine;
  /** Ordered op log — tests assert the lifecycle order against this. */
  operations: string[];
  screenshotPath: string;
  screenshotBytes: number;
  terminalResult: Record<string, unknown>;
  fileContent: Record<string, unknown>;
  browserResult: Record<string, unknown>;
  snapshotId: string;
  /** Sum of X-Credits-Charged across all calls (0 on sandbox keys). */
  creditsObserved: number;
  /** cost.ts estimate for the elapsed runtime + the snapshot. */
  estimatedCredits: number;
  elapsedMinutes: number;
}

export const REMOTE_FILE_PATH = '/tmp/coasty-cookbook-ex08.txt';
export const REMOTE_FILE_CONTENT = 'hello from the coasty cookbook (ex08)';

async function pollUntilRunning(
  client: CoastyClient,
  machineId: string,
  options: { sleep: SleepFn; intervalMs: number; maxPolls: number },
): Promise<Machine> {
  for (let poll = 0; poll < options.maxPolls; poll += 1) {
    const { data: machine } = await client.machines.get(machineId);
    if (machine.status === 'running') return machine;
    if (machine.status === 'error' || machine.status === 'terminated') {
      throw new MachineOpFailedError('poll', `machine entered status "${machine.status}"`, null);
    }
    await options.sleep(options.intervalMs);
  }
  throw new MachineOpFailedError(
    'poll',
    `machine not running after ${String(options.maxPolls)} polls`,
    null,
  );
}

/**
 * The full lifecycle. The machine is ALWAYS stopped + terminated in `finally`
 * once provisioned — even when a mid-lifecycle call throws.
 */
export async function runMachineLifecycle(
  client: CoastyClient,
  options: MachineExampleOptions,
): Promise<MachineExampleReport> {
  const log = options.logger ?? ((line: string): void => void process.stdout.write(`${line}\n`));
  const osType = options.osType ?? 'linux';
  const ttlMinutes = options.ttlMinutes ?? DEFAULT_TTL_MINUTES;
  const sleep = options.sleep ?? defaultSleep;
  const now = options.now ?? ((): number => Date.now());

  // Guards FIRST — nothing is provisioned unless both pass.
  validateTtlMinutes(ttlMinutes);
  if (!options.sandbox && !options.confirmedSpend) throw new SpendNotConfirmedError(osType);

  const hourly = machineHourlyCredits(osType, 'running');
  log(
    options.sandbox
      ? 'sandbox key: instant free mch_test_* VM, no wallet gate, $0.00'
      : `live key: gate wallet >= ${String(PRICING.provisioningGateCredits)} cr ($0.20, not a fee); ` +
          `${osType} runs at ${String(hourly)} cr/hr (${String(PRICING.machineHourly.stopped)} cr/hr stopped)`,
  );

  const operations: string[] = [];
  let creditsObserved = 0;
  const track = <T extends { meta: { creditsCharged: number | null } }>(result: T): T => {
    creditsObserved += result.meta.creditsCharged ?? 0;
    return result;
  };

  const startedAtMs = now();

  // 1. Provision with the auto-terminate TTL.
  const provisioned = track(
    await client.machines.provision(
      {
        display_name: options.displayName ?? 'cookbook-ex08',
        os_type: osType,
        desktop_enabled: true,
        ttl_minutes: ttlMinutes,
      },
      options.idempotencyKey === undefined ? {} : { idempotencyKey: options.idempotencyKey },
    ),
  );
  operations.push('provision');
  let machine = provisioned.data.machine;
  log(`provisioned ${machine.id} (status ${machine.status}, ttl ${String(ttlMinutes)} min)`);

  try {
    // 2. Poll until running (sandbox machines are usually running instantly).
    if (machine.status !== 'running') {
      machine = await pollUntilRunning(client, machine.id, {
        sleep,
        intervalMs: options.pollIntervalMs ?? 2000,
        maxPolls: options.maxPolls ?? 60,
      });
    } else {
      // Still record one poll so the lifecycle order is deterministic.
      machine = track(await client.machines.get(machine.id)).data;
    }
    operations.push('poll');
    log(`machine ${machine.id} is running`);

    // 3. Screenshot -> save to disk (free per call).
    const shot = track(await client.machines.screenshot(machine.id));
    const imageBytes = Buffer.from(shot.data.image_b64, 'base64');
    await writeFile(options.screenshotPath, imageBytes);
    operations.push('screenshot');
    log(
      `saved screenshot (${String(shot.data.width)}x${String(shot.data.height)}, ` +
        `${String(imageBytes.length)} bytes) -> ${options.screenshotPath}`,
    );

    // 4. One low-level action (free).
    const action = track(
      await client.machines.action(machine.id, {
        command: 'click',
        parameters: { x: 640, y: 360 },
      }),
    );
    operations.push('action');
    if (!action.data.success) {
      throw new MachineOpFailedError('actions:click', action.data.error, action.meta.requestId);
    }

    // 5. A batched action sequence (<= 50 steps, stop_on_error).
    const batch = track(
      await client.machines.actionsBatch(machine.id, {
        steps: [
          { command: 'move', parameters: { x: 320, y: 240 } },
          { command: 'type_text', parameters: { text: 'coasty cookbook' } },
          { command: 'key_press', parameters: { key: 'enter' } },
        ],
        stop_on_error: true,
      }),
    );
    operations.push('actions_batch');
    if (batch.data.failed_count > 0 || batch.data.aborted) {
      throw new MachineOpFailedError(
        'actions/batch',
        `${String(batch.data.failed_count)} step(s) failed (aborted=${String(batch.data.aborted)})`,
        batch.meta.requestId,
      );
    }

    // 6. Terminal (bash on Linux, PowerShell on Windows; free).
    const terminal = track(
      await client.machines.terminal(machine.id, {
        command: "echo 'hello from the coasty cookbook'",
        timeout_ms: 30000,
      }),
    );
    operations.push('terminal');

    // 7. Files: write then read it back (free).
    track(
      await client.machines.files(machine.id, 'write', {
        path: REMOTE_FILE_PATH,
        content: REMOTE_FILE_CONTENT,
      }),
    );
    operations.push('file_write');
    const fileRead = track(
      await client.machines.files(machine.id, 'read', { path: REMOTE_FILE_PATH }),
    );
    operations.push('file_read');

    // 8. Browser navigate (free).
    const browser = track(
      await client.machines.browser(machine.id, 'navigate', {
        parameters: { url: 'https://example.com' },
      }),
    );
    operations.push('browser_navigate');

    // 9. Snapshot — 1 credit ($0.01) one-time, refunded on failure.
    const snapshot = track(
      await client.machines.snapshot(
        machine.id,
        options.idempotencyKey === undefined
          ? {}
          : { idempotencyKey: `${options.idempotencyKey}-snap` },
      ),
    );
    operations.push('snapshot');
    log(
      `snapshot ${snapshot.data.snapshot_id} (${String(snapshot.data.credits_charged)} cr charged)`,
    );

    const elapsedMinutes = (now() - startedAtMs) / 60_000;
    const estimatedCredits =
      estimateMachineRuntimeCredits({ osType, minutes: elapsedMinutes }) + snapshotCredits();
    log(
      formatEstimate(
        [
          {
            label: `runtime ${elapsedMinutes.toFixed(1)} min @ ${String(hourly)} cr/hr (floored)`,
            credits: estimateMachineRuntimeCredits({ osType, minutes: elapsedMinutes }),
          },
          { label: 'snapshot (one-time)', credits: snapshotCredits() },
          { label: 'per-call ops (actions/terminal/files/browser/screenshot)', credits: 0 },
        ],
        { sandbox: options.sandbox },
      ),
    );
    log(`credits actually observed via X-Credits-Charged: ${String(creditsObserved)}`);

    return {
      machine,
      operations,
      screenshotPath: options.screenshotPath,
      screenshotBytes: imageBytes.length,
      terminalResult: terminal.data,
      fileContent: fileRead.data,
      browserResult: browser.data,
      snapshotId: snapshot.data.snapshot_id,
      creditsObserved,
      estimatedCredits,
      elapsedMinutes,
    };
  } finally {
    // 10. ALWAYS stop (drops to the 1 cr/hr storage rate) then terminate
    //     (ends all billing) — even when the body threw. Cleanup failures are
    //     reported loudly but never mask the original error.
    for (const [op, call] of [
      ['stop', (): Promise<unknown> => client.machines.stop(machine.id)],
      ['terminate', (): Promise<unknown> => client.machines.terminate(machine.id)],
    ] as const) {
      try {
        await call();
        operations.push(op);
        log(`${op}: ok (${machine.id})`);
      } catch (cleanupError) {
        const requestId = cleanupError instanceof CoastyError ? cleanupError.requestId : null;
        console.error(
          `cleanup "${op}" failed for ${machine.id}: ${String(cleanupError)} ` +
            `(request_id: ${requestId ?? 'n/a'}) — the ${String(ttlMinutes)}-min TTL is the backstop`,
        );
      }
    }
  }
}

// ---------------------------------------------------------------------------
// Thin CLI
// ---------------------------------------------------------------------------

function flagValue(argv: string[], flag: string): string | undefined {
  const index = argv.indexOf(flag);
  return index === -1 ? undefined : argv[index + 1];
}

function isMain(): boolean {
  const entry = process.argv[1];
  return entry !== undefined && path.resolve(entry) === fileURLToPath(import.meta.url);
}

export async function main(argv: string[] = process.argv.slice(2)): Promise<void> {
  const print = (line: string): void => void process.stdout.write(`${line}\n`);
  const sandbox = isSandboxKey();
  const osFlag = flagValue(argv, '--os') ?? 'linux';
  if (osFlag !== 'linux' && osFlag !== 'windows') {
    console.error(`--os must be linux or windows (got ${osFlag})`);
    process.exitCode = 1;
    return;
  }
  const ttlMinutes = Number(flagValue(argv, '--ttl') ?? String(DEFAULT_TTL_MINUTES));
  const screenshotPath = flagValue(argv, '--out') ?? path.resolve('ex08-screenshot.png');

  print(
    formatEstimate(
      [
        {
          label: `runtime <= ${String(ttlMinutes)} min @ ${String(machineHourlyCredits(osFlag, 'running'))} cr/hr`,
          credits: estimateMachineRuntimeCredits({ osType: osFlag, minutes: ttlMinutes }),
        },
        { label: 'snapshot (one-time, refunded on failure)', credits: snapshotCredits() },
      ],
      { sandbox },
    ),
  );
  if (!sandbox) {
    print(
      `note: live provisioning also requires wallet >= ${String(PRICING.provisioningGateCredits)} cr ` +
        `($${creditsToUsd(PRICING.provisioningGateCredits).toFixed(2)}) — a gate, not a fee.`,
    );
  }

  const client = new CoastyClient();
  const report = await runMachineLifecycle(client, {
    sandbox,
    confirmedSpend: argv.includes('--confirm') || spendConfirmed(),
    screenshotPath,
    osType: osFlag,
    ttlMinutes,
  });
  print(`done: ${report.operations.join(' -> ')}`);
}

if (isMain()) {
  main().catch((error: unknown) => {
    if (error instanceof CoastyError) {
      console.error(`[${error.code}] ${error.message} (request_id: ${error.requestId ?? 'n/a'})`);
    } else {
      console.error(String(error));
    }
    process.exitCode = 1;
  });
}
